crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs

```diff
@@ -0,0 +1,529 @@
+                    if chunk_index == self.chunks.len() - 1 {
+                        diffs.push(VectorDiff::Append { values: items.into() });
+                    }
+                    // No optimisation: let's emit `VectorDiff::Insert`.
+                    else {
+                        diffs.extend(items.into_iter().enumerate().map(|(nth, item)| {
+                            VectorDiff::Insert { index: offset + nth, value: item.clone() }
+                        }));
+                    }
+                }
+
+                Update::DetachLastItems { at } => {
+                    let expected_chunk_identifier = at.chunk_identifier();
+                    let new_length = at.index();
+
+                    let length = self
+                        .chunks
+                        .iter_mut()
+                        .find_map(|(chunk_identifier, length)| {
+                            (*chunk_identifier == expected_chunk_identifier).then_some(length)
+                        })
+                        // SAFETY: Assuming `LinkedChunk` and `Updates` are not buggy, and assuming
+                        // `Self::chunks` is correctly initialized, it is not possible to detach
+                        // items from a chunk that does not exist. If this predicate fails, it means
+                        // `LinkedChunk` or `Updates` contain a bug.
+                        .expect("Detach last items: The chunk is not found");
+
+                    *length = new_length;
+                }
+
+                Update::ReattachItems => {
+                    // Entering the `reattaching` mode.
+                    mute_push_items = true;
+                }
+
+                Update::ReattachItemsDone => {
+                    // Exiting the `reattaching` mode.
+                    mute_push_items = false;
+                }
+            }
+        }
+
+        diffs
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use imbl::vector;
+
+    use super::{super::LinkedChunk, VectorDiff};
+
+    #[test]
+    fn test_as_vector() {
+        let mut linked_chunk = LinkedChunk::<3, char, ()>::new_with_update_history();
+        let mut as_vector = linked_chunk.as_vector().unwrap();
+
+        assert!(as_vector.take().is_empty());
+
+        linked_chunk.push_items_back(['a', 'b', 'c', 'd']);
+        #[rustfmt::skip]
+        assert_items_eq!(linked_chunk, ['a', 'b', 'c'] ['d']);
+
+        // From an `ObservableVector` point of view, it would look like:
+        //
+        // 0   1   2   3   4
+        // +---+---+---+---+
+        // | a | b | c | d |
+        // +---+---+---+---+
+        // ^^^^^^^^^^^^^^^^
+        // |
+        // new
+        assert_eq!(
+            as_vector.take(),
+            &[
+                VectorDiff::Append { values: vector!['a', 'b', 'c'] },
+                VectorDiff::Append { values: vector!['d'] },
+            ]
+        );
```

Comment:
Nice! Alternatively, or in addition to those checks, could we also have the corresponding vector that we're building, and applying the diffs then checking its content? Then the comments don't really need to be put there, we can "just" have the assertions on the entire vectors:

```
assert_eq!(reconstructed_vector, vector!['a', 'b', 'c', 'd']);
```

Reply:
Done.